librz/core/cmd/cmd_flag.c

```diff
@@ -516,100 +504,80 @@ static void cmd_flag_table(RzCore *core, const char *input) {
 	rz_table_free(t);
 }
 
-static void cmd_flag_tags(RzCore *core, const char *input) {
-	char mode = input[1];
-	for (; *input && !IS_WHITESPACE(*input); input++) {
-	}
-	char *inp = strdup(input);
-	char *arg = inp;
-	rz_str_trim(arg);
-	if (!*arg && !mode) {
-		const char *tag;
-		RzListIter *iter;
-		RzList *list = rz_flag_tags_list(core->flags, NULL);
-		rz_list_foreach (list, iter, tag) {
-			rz_cons_printf("%s\n", tag);
-		}
-		rz_list_free(list);
-		free(inp);
-		return;
-	}
-	if (mode == '?') {
-		rz_core_cmd_help(core, help_msg_ft);
-		free(inp);
-		return;
-	}
-	if (mode == 'w') { // "ftw"
-		const char *tag;
-		RzListIter *iter;
-		RzList *list = rz_flag_tags_list(core->flags, NULL);
-		rz_list_foreach (list, iter, tag) {
-			rz_cons_printf("%s:\n", tag);
-			rz_core_cmdf(core, "ftn %s", tag);
+RZ_IPI RzCmdStatus rz_flag_tag_add_handler(RzCore *core, int argc, const char **argv) {
+	rz_flag_tags_set(core->flags, argv[1], argv[2]);
+	return RZ_CMD_STATUS_OK;
+}
+
+static void flag_tag_print(RzCore *core, const char *tag, RzCmdStateOutput *state) {
+	PJ *pj = state->d.pj;
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_JSON: {
+		pj_k(pj, tag);
+		pj_a(pj);
+		RzList *flags = rz_flag_tags_get(core->flags, tag);
+		if (!flags) {
+			pj_end(pj);
+			break;
 		}
-		rz_list_free(list);
-		free(inp);
-		return;
-	}
-	if (mode == '*') { // "ft*"
 		RzListIter *iter;
-		const char *tag;
-		RzList *list = rz_flag_tags_list(core->flags, NULL);
-		rz_list_foreach (list, iter, tag) {
-			const char *flags = sdb_get(core->flags->tags, sdb_fmt("tag.%s", tag), NULL);
-			rz_cons_printf("ft %s %s\n", tag, flags);
-		}
-		rz_list_free(list);
-		free(inp);
-		return;
-	}
-	if (mode == 'j') { // "ftj"
-		RzListIter *iter, *iter2;
-		const char *tag, *flg;
-		PJ *pj = pj_new();
-		pj_o(pj);
-		RzList *list = rz_flag_tags_list(core->flags, NULL);
-		rz_list_foreach (list, iter, tag) {
-			pj_k(pj, tag);
-			pj_a(pj);
-			RzList *flags = rz_flag_tags_list(core->flags, tag);
-			rz_list_foreach (flags, iter2, flg) {
-				pj_s(pj, flg);
-			}
-			pj_end(pj);
-			rz_list_free(flags);
+		RzFlagItem *flag;
+		rz_list_foreach (flags, iter, flag) {
+			pj_s(pj, flag->name);
 		}
 		pj_end(pj);
-		rz_list_free(list);
-		free(inp);
-		rz_cons_printf("%s\n", pj_string(pj));
-		pj_free(pj);
-		return;
+		rz_list_free(flags);
+		break;
 	}
-	char *arg1 = strchr(arg, ' ');
-	if (arg1) {
-		*arg1 = 0;
-		const char *a1 = rz_str_trim_head_ro(arg1 + 1);
-		rz_flag_tags_set(core->flags, arg, a1);
-	} else {
+	case RZ_OUTPUT_MODE_LONG: {
+		rz_cons_printf("%s:\n", tag);
+		RzList *flags = rz_flag_tags_get(core->flags, tag);
+		if (!flags) {
+			break;
+		}
```

Comment:
you don't really need to check for NULL here. rz_list_foreach does that and so does the free method

Reply:
I think it's clearer this way.